### PR TITLE
[dagster-airlift] complete revamp

### PR DESCRIPTION
**Pull request by dpeng817**
Complete project revamp.

### CODE REVIEW

**Comment by dpeng817** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/23369?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#23370** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/23370?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#23369** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/23369?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#23344** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/23344?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#23343** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/23343?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#23326** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/23326?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#23324** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/23324?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#23300** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/23300?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @dpeng817 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: examples/experimental/dagster-airlift/dagster_airlift/airflow_utils.py [hunk 1]
```diff
@@ -390,3 +395,160 @@ def toposort_specs(specs: List[AssetSpec]) -> List[AssetSpec]:
...
+# / blueprints
+#   / dag_id
+#     / task_id.yaml
+# Each task_id.yaml will look something like this:
+# - asset: my/asset/key
+#   tags: {}
+#   metadata: {}
+#   deps: [upstream/asset/key]
+#   compute_pointer: my.compute.pointer
+# - asset: my/other/asset/key
```
Comment (schrockn) [inline examples/experimental/dagster-airlift/dagster_airlift/airflow_utils.py:410]: this is totally configurable on a per-blueprint basis, correct?

File: examples/experimental/dagster-airlift/dagster_airlift_tests/airflow_migration_project/migrated.yaml [hunk 2]
```diff
@@ -0,0 +1,5 @@
+dags:
```
Comment (schrockn) [inline examples/experimental/dagster-airlift/dagster_airlift_tests/airflow_migration_project/migrated.yaml:1]: can easily change this later but I do think we will want one file per dag

File: examples/experimental/dagster-airlift/dagster_airlift/airflow_utils.py [hunk 3]
```diff
@@ -390,3 +395,160 @@ def toposort_specs(specs: List[AssetSpec]) -> List[AssetSpec]:
...
+                    module = import_module(ptr.module_name)
+                    compute_fn = getattr(module, ptr.fn_name)
+                    compute_fn()
+
+        return Definitions(assets=[_external_asset])
+
+
+class TaskMigrationState(BaseModel):
+    task_id: str
+    migrated: bool
```
Comment (schrockn) [inline examples/experimental/dagster-airlift/dagster_airlift/airflow_utils.py:472]: For our own sake, I think we should split these into different files, and make it super clear what we expect to be in the airflow process versus the dagster process. We will probably want to have different modules eventually.

File: examples/experimental/dagster-airlift/dagster_airlift/airflow_utils.py [hunk 4]
```diff
@@ -390,3 +395,160 @@ def toposort_specs(specs: List[AssetSpec]) -> List[AssetSpec]:
...
+        )
+        def _external_asset() -> None:
+            if self.migrated:
+                for info in self.asset_infos:
+                    ptr = check.not_none(info.asset_compute_pointer)
+                    module = import_module(ptr.module_name)
+                    compute_fn = getattr(module, ptr.fn_name)
+                    compute_fn()
+
+        return Definitions(assets=[_external_asset])
```
Comment (schrockn) [inline examples/experimental/dagster-airlift/dagster_airlift/airflow_utils.py:467]: I'd also like to see the pattern where have an existing blueprint (e.g. dbt) and have it only return specs when it is not migrated.

**Review by schrockn — CHANGES_REQUESTED** [review]
This one probably worth talking through live.

File: examples/experimental/dagster-airlift/dagster_airlift/airflow_utils.py (hunk 1 above)
Comment (schrockn) [inline examples/experimental/dagster-airlift/dagster_airlift/airflow_utils.py:410]:
```yaml
type: python
python_fn: my_module.python_fn
assets:
  - key:
     tags:
     metadata:
     deps:
```

File: examples/experimental/dagster-airlift/dagster_airlift/airflow_utils.py [hunk 5]
```diff
@@ -390,3 +395,160 @@ def toposort_specs(specs: List[AssetSpec]) -> List[AssetSpec]:
...
+            name=self.name,
+            compute_kind=self.compute_kind,
+        )
+        def _external_asset() -> None:
+            if self.migrated:
+                for info in self.asset_infos:
+                    ptr = check.not_none(info.asset_compute_pointer)
+                    module = import_module(ptr.module_name)
+                    compute_fn = getattr(module, ptr.fn_name)
+                    compute_fn()
```
Comment (schrockn) [inline examples/experimental/dagster-airlift/dagster_airlift/airflow_utils.py:465]: accessing the attr should happen in the factory function, not the body of the multi-asset

**Review by schrockn — COMMENTED** [review]
Flushing through comments

File: examples/experimental/dagster-airlift/dagster_airlift/airflow_utils.py (hunk 3 above)
Comment (dpeng817) [reply examples/experimental/dagster-airlift/dagster_airlift/airflow_utils.py:472]: Heavily agreed about the different modules. I figure for a lot of people adding a heavyweight dependency to dagster within their airflow code will be a non-starter. I've been thinking that airlift will consist of the dagster component, and then a lightweight-pipes esque package with the only dependency being requests. Maybe we could even get away without that.

I'll put them into their own file for now.

File: examples/experimental/dagster-airlift/dagster_airlift/dbt/multi_asset.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+class DbtProjectDefs(Blueprint):
+    type: Literal["dbt_project"] = "dbt_project"
+    dbt_project_path: FromEnvVar
+    group: Optional[str] = None
+
+    def build_defs(self) -> Definitions:
+        source_file_name_without_ext = self.source_file_name.split(".")[0]
+        dbt_project_path = Path(os.environ[self.dbt_project_path.env_var])
+        dbt_manifest_path = dbt_project_path / "target" / "manifest.json"
+
```
Comment (schrockn) [inline examples/experimental/dagster-airlift/dagster_airlift/dbt/multi_asset.py:25]: Would prefer to see `FromEnvVar` go away so we don't rely on it

File: examples/experimental/dagster-airlift/examples/peering-with-dbt/tox.ini [hunk 7]
```diff
@@ -10,9 +10,14 @@ deps =
...
   -e ../../../../../python_modules/dagster-test
   -e ../../../../../python_modules/dagster-pipes
-  -e ../../../../../python_modules/libraries/dagster-dbt
   -e ../../../../../python_modules/dagster-graphql
-  -e ../../../dagster-airlift
+  -e ../../../../../python_modules/libraries/dagster-dbt
+  -e ../../../../../python_modules/libraries/dagster-databricks
+  -e ../../../../../python_modules/libraries/dagster-pandas
+  -e ../../../../../python_modules/libraries/dagster-pyspark
+  -e ../../../../../python_modules/libraries/dagster-spark
```
Comment (schrockn) [inline examples/experimental/dagster-airlift/examples/peering-with-dbt/tox.ini:18]: why do we need all of these?

**Review by schrockn — APPROVED** [review]
Ok given that this is experimental we can fix forward stuff, but

1) Eliminating unnecessary dependencies very important (e.g. early design partners who do not use databricks should not install dagster-databricks)
2) We should cut our dependency on blueprints.

**Pull request merged**